mem-store.ts

```diff
@@ -2,12 +2,209 @@
 'use strict'
 
 import Assert from 'assert'
-import * as Common from './lib/common'
 
 let internals = {
   name: 'mem-store'
 }
 
+const intern = {
+  /* NOTE: `intern` serves as a namespace for utility functions used by
+   * the mem store.
+   */
+
+
+  // NOTE: This function is intended for use by the #save method. This
+  // function returns true when the entity argument is assumed to not yet
+  // exist in the store.
+  //
+  // In terms of code, if client code looks like so:
+  // ```
+  //   seneca.make('product')
+  //     .data$({ label, price })
+  //     .save$(done)
+  // ```
+  //
+  // - `intern.is_new` will be invoked from the #save method and return
+  // true, because the product entity is yet to be saved.
+  //
+  // The following client code will cause `intern.is_new` to return false,
+  // when invoked from the #save method, because the user entity already
+  // exists:
+  // ```
+  //   seneca.make('user')
+  //     .load$(user_id, (err, user) => {
+  //       if (err) return done(err)
+  //
+  //       return user
+  //         .data$({ email, username })
+  //         .save$(done)
+  //     })
+  // ```
+  // 
+  is_new(ent: any): boolean {
+    return null != ent && null == ent.id
+  },
+
+
+  is_upsert_requested(msg: any): boolean {
+    const { ent, q } = msg
+    return intern.is_new(ent) && Array.isArray(q.upsert$)
+  },
+
+
+  find_one_doc(entmap: any, ent: any, filter: any): any {
+    const { base, name } = ent.canon$({ object: true })
+
+    if (!(base in entmap)) {
+      return null
+    }
+
+    if (!(name in entmap[base])) {
+      return null
+    }
+
+    const coll = entmap[base][name]
+    const docs = Object.values(coll)
+    const public_entdata = ent.data$(false) 
+
+    const doc = docs.find((doc: any) => {
```

Comment:
I don't think it would be appropriate in this scenario, because `doc` refers to raw memory. Such document can then be cast into entity:
```
ent.make$(doc)
```

Reply:
Oh, please disregard my last message, I see now.